--- multipath-tools/multipathd/main.c	2009-04-01 14:47:15.000000000 -0400
+++ multipath-tools/multipathd/main.c	2009-04-01 15:22:43.000000000 -0400
@@ -1186,6 +1186,7 @@
 	add_handler(DISABLEQ+MAP, cli_disable_queueing);
 	add_handler(RESTOREQ+MAP, cli_restore_queueing);
 
+	umask(077);
 	uxsock_listen(&uxsock_trigger, ap);
 
 	return NULL;
@@ -1759,7 +1760,6 @@
 	close(in_fd);
 	close(out_fd);
 	chdir("/");
-	umask(0);
 	return 0;
 }
 
