--- multipath-tools-0.4.8/multipath.conf.annotated~	2009-10-26 16:56:24.000000000 +0200
+++ multipath-tools-0.4.8/multipath.conf.annotated	2009-10-26 16:56:57.028914153 +0200
@@ -7,7 +7,7 @@
 ## name  : defaults
 ## desc  : multipath-tools default settings
 ##
-#defaults {
+defaults {
 #	#
 #	# name    : udev_dir
 #	# desc    : directory where udev creates its device nodes
@@ -163,8 +163,8 @@
 #	#           file.
 #	# values  : yes|no
 #	# default : no
-#	user_friendly_names no
-#
+	user_friendly_names yes
+
 #	#
 #	# name    : mode
 #	# scope   : multipath
@@ -191,8 +191,8 @@
 #	# default : determined by the process
 #	gid disk
 #
-#}
-#	
+}
+
 ##
 ## name    : blacklist
 ## scope   : multipath & multipathd
