--- multipath-tools-0.4.9/kpartx/kpartx.rules~	2013-05-17 23:53:47.000000000 +0300
+++ multipath-tools-0.4.9/kpartx/kpartx.rules	2013-05-17 23:54:09.123094293 +0300
@@ -7,6 +7,9 @@
 KERNEL!="dm-*", GOTO="kpartx_end"
 ACTION=="remove", GOTO="kpartx_end"
 
+IMPORT{program}="/sbin/dmsetup export -j%M -m%m"
+ENV{DM_NAME}!="?*", GOTO="kpartx_end"
+
 ENV{DM_TABLE_STATE}!="LIVE", GOTO="kpartx_end"
 
 ENV{DM_UUID}=="?*", IMPORT{program}=="/lib/udev/kpartx_id %M %m $env{DM_UUID}"
@@ -36,7 +36,7 @@
 
 # Create dm tables for partitions
 ENV{DM_STATE}!="SUSPENDED", ENV{DM_UUID}=="mpath-*", \
-        RUN+="/sbin/kpartx -u -p -part /dev/$name"
+        RUN+="/sbin/kpartx -u /dev/$name"
 
 LABEL="kpartx_end"
 
