--- multipath-tools-0.5.0/libmultipath/checkers/rdac.c.orig	2013-12-17 22:40:41.000000000 +0100
+++ multipath-tools-0.5.0/libmultipath/checkers/rdac.c	2014-02-16 21:33:01.573751205 +0100
@@ -308,7 +308,7 @@
 done:
 	switch (ret) {
 	case PATH_DOWN:
-		MSG(c, (inqfail) ? MSG_RDAC_DOWN_TYPE("inquiry failed") :
+		MSG(c, "%s", (inqfail) ? MSG_RDAC_DOWN_TYPE("inquiry failed") :
 			checker_msg_string(&inq));
 		break;
 	case PATH_UP:
