--- multipath-tools-0.4.7-o/devmap_name/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/devmap_name/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -10,7 +10,7 @@
 ifeq ($(strip $(BUILD)),klibc)
 	OBJS += $(libdm)
 else
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 endif
 
 EXEC = devmap_name
--- multipath-tools-0.4.7-o/kpartx/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/kpartx/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -13,7 +13,7 @@
 	       lopart.o xstrncpy.o devmapper.o dasd.o mac.o \
 	       $(MULTIPATHLIB)-$(BUILD).a $(libdm)
 else
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 	OBJS = bsd.o dos.o kpartx.o solaris.o unixware.o dasd.o \
 	       gpt.o mac.o crc32.o lopart.o xstrncpy.o devmapper.o
 endif
--- multipath-tools-0.4.7-o/multipath/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/multipath/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -12,7 +12,7 @@
 ifeq ($(strip $(BUILD)),klibc)
 	OBJS += $(libdm) $(libsysfs)
 else
-	LDFLAGS += -ldevmapper -lsysfs
+	LDFLAGS += -ldevmapper -lselinux -lsepol -lsysfs
 endif
 
 EXEC = multipath
--- multipath-tools-0.4.7-o/multipathd/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/multipathd/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -7,7 +7,7 @@
 # basic flags setting
 #
 CFLAGS += -DDAEMON -I$(multipathdir) -I$(checkersdir)
-LDFLAGS = -lpthread -ldevmapper -lsysfs -lreadline -lncurses
+LDFLAGS = -lpthread -ldevmapper -lselinux -lsepol -lsysfs -lreadline -lncurses
 
 #
 # debuging stuff
--- multipath-tools-0.4.7-o/path_priority/pp_balance_units/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/path_priority/pp_balance_units/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -13,7 +13,7 @@
 	OBJS = pp_balance_units.o $(MULTIPATHLIB)-$(BUILD).a
 else
 	CFLAGS += -I$(multipathdir) -DDEBUG=$(DEBUG)
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 	OBJS = pp_balance_units.o $(MULTIPATHLIB)-$(BUILD).a
 endif
 
