diff -ur 1.org/libmultipath/log.c 1/libmultipath/log.c
--- 1.org/libmultipath/log.c	2010-05-22 14:01:58.000000000 +0200
+++ 1/libmultipath/log.c	2010-11-27 19:14:10.979257334 +0100
@@ -142,7 +142,7 @@
 	la->empty = 0;
 	msg = (struct logmsg *)la->tail;
 	msg->prio = prio;
-	strcpy((void *)&msg->str, buff);
+	memcpy((void *)&msg->str, buff, strlen(buff) + 1);
 	lastmsg->next = la->tail;
 	msg->next = la->head;
 
