--- multipath-tools-0.4.7-o/devmap_name/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/devmap_name/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -10,7 +10,7 @@
 ifeq ($(strip $(BUILD)),klibc)
 	OBJS += $(libdm)
 else
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 endif
 
 EXEC = devmap_name
--- multipath-tools-0.4.8/kpartx/Makefile.orig	2007-08-02 23:05:37.000000000 +0200
+++ multipath-tools-0.4.8/kpartx/Makefile	2007-09-10 00:28:52.386238483 +0200
@@ -13,7 +13,7 @@
 	       lopart.o xstrncpy.o devmapper.o dasd.o mac.o sun.o \
 	       $(MULTIPATHLIB)-$(BUILD).a $(libdm)
 else
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 	OBJS = bsd.o dos.o kpartx.o solaris.o unixware.o dasd.o sun.o \
 	       gpt.o mac.o crc32.o lopart.o xstrncpy.o devmapper.o
 endif
--- multipath-tools-0.4.8/multipath/Makefile.orig	2007-08-02 23:05:37.000000000 +0200
+++ multipath-tools-0.4.8/multipath/Makefile	2007-09-10 00:29:21.867918548 +0200
@@ -13,7 +13,7 @@
 ifeq ($(strip $(BUILD)),klibc)
 	OBJS += $(libdm)
 else
-	LDFLAGS += -ldevmapper
+	LDFLAGS += -ldevmapper -lselinux -lsepol
 endif
 
 EXEC = multipath
--- multipath-tools-0.4.8/multipathd/Makefile.orig	2007-08-02 23:05:37.000000000 +0200
+++ multipath-tools-0.4.8/multipathd/Makefile	2007-09-10 00:29:51.665616622 +0200
@@ -7,7 +7,7 @@
 # basic flags setting
 #
 CFLAGS += -DDAEMON -I$(multipathdir) -I$(checkersdir)
-LDFLAGS = -lpthread -ldevmapper -lreadline -lncurses -laio
+LDFLAGS = -lpthread -ldevmapper -lselinux -lsepol -lreadline -lncurses -laio
 
 #
 # debuging stuff
--- multipath-tools-0.4.7-o/path_priority/pp_balance_units/Makefile	2006-03-13 04:07:45.000000000 -0700
+++ multipath-tools-0.4.7/path_priority/pp_balance_units/Makefile	2006-07-29 04:11:48.000000000 -0600
@@ -13,7 +13,7 @@
 	OBJS = pp_balance_units.o $(MULTIPATHLIB)-$(BUILD).a
 else
 	CFLAGS += -I$(multipathdir) -DDEBUG=$(DEBUG)
-	LDFLAGS = -ldevmapper
+	LDFLAGS = -ldevmapper -lselinux -lsepol
 	OBJS = pp_balance_units.o $(MULTIPATHLIB)-$(BUILD).a
 endif
 
