--- multipath-tools-0.8.0-eb688e1/Makefile.inc.orig	2019-02-14 18:55:29.000000000 +0100
+++ multipath-tools-0.8.0-eb688e1/Makefile.inc	2019-03-10 14:10:20.140218925 +0100
@@ -36,8 +36,8 @@
 endif
 
 ifndef SYSTEMD
-	ifeq ($(shell systemctl --version > /dev/null 2>&1 && echo 1), 1)
-		SYSTEMD = $(shell systemctl --version 2> /dev/null |  sed -n 's/systemd \([0-9]*\)/\1/p')
+	ifeq ($(shell pkg-config --exists systemd > /dev/null 2>&1 && echo 1), 1)
+		SYSTEMD = $(shell  pkg-config --modversion systemd 2> /dev/null)
 	endif
 endif
 
