--- ./kpartx/lopart.c~	2005-09-02 16:14:18.000000000 +0300
+++ ./kpartx/lopart.c	2005-10-16 03:44:40.000000000 +0300
@@ -27,15 +27,6 @@
 #include <sys/mman.h>
 #include <sysmacros.h>
 
-#if defined(__hppa__) || defined(__powerpc64__) || defined (__alpha__) \
- || defined (__x86_64__)
-typedef unsigned long __kernel_old_dev_t;
-#elif defined(__powerpc__) || defined(__ia64__) || (defined(__sparc__) && defined (__arch64__))
-typedef unsigned int __kernel_old_dev_t;
-#else
-typedef unsigned short __kernel_old_dev_t;
-#endif
-
 #define dev_t __kernel_old_dev_t
 
 #include <linux/loop.h>
